common/client/transaction_sender.go

```diff
@@ -0,0 +1,278 @@
+	chainID           CHAIN_ID
+	chainFamily       string
+	lggr              logger.SugaredLogger
+	multiNode         *MultiNode[CHAIN_ID, RPC]
+	txErrorClassifier TxErrorClassifier[TX]
+	sendTxSoftTimeout time.Duration // defines max waiting time from first response til responses evaluation
+
+	wg     sync.WaitGroup // waits for all reporting goroutines to finish
+	chStop services.StopChan
+}
+
+// SendTransaction - broadcasts transaction to all the send-only and primary nodes in MultiNode.
+// A returned nil or error does not guarantee that the transaction will or won't be included. Additional checks must be
+// performed to determine the final state.
+//
+// Send-only nodes' results are ignored as they tend to return false-positive responses. Broadcast to them is necessary
+// to speed up the propagation of TX in the network.
+//
+// Handling of primary nodes' results consists of collection and aggregation.
+// In the collection step, we gather as many results as possible while minimizing waiting time. This operation succeeds
+// on one of the following conditions:
+// * Received at least one success
+// * Received at least one result and `sendTxSoftTimeout` expired
+// * Received results from the sufficient number of nodes defined by sendTxQuorum.
+// The aggregation is based on the following conditions:
+// * If there is at least one success - returns success
+// * If there is at least one terminal error - returns terminal error
+// * If there is both success and terminal error - returns success and reports invariant violation
+// * Otherwise, returns any (effectively random) of the errors.
+func (txSender *TransactionSender[TX, CHAIN_ID, RPC]) SendTransaction(ctx context.Context, tx TX) (SendTxReturnCode, error) {
+	txResults := make(chan sendTxResult)
+	txResultsToReport := make(chan sendTxResult)
+	primaryNodeWg := sync.WaitGroup{}
+
+	ctx, cancel := txSender.chStop.Ctx(ctx)
+	defer cancel()
+
+	healthyNodesNum := 0
+	err := txSender.multiNode.DoAll(ctx, func(ctx context.Context, rpc RPC, isSendOnly bool) {
+		if isSendOnly {
+			txSender.wg.Add(1)
+			go func() {
+				defer txSender.wg.Done()
+				// Send-only nodes' results are ignored as they tend to return false-positive responses.
+				// Broadcast to them is necessary to speed up the propagation of TX in the network.
+				_ = txSender.broadcastTxAsync(ctx, rpc, tx)
+			}()
+			return
+		}
+
+		// Primary Nodes
+		healthyNodesNum++
+		primaryNodeWg.Add(1)
+		go func() {
+			defer primaryNodeWg.Done()
+			result := txSender.broadcastTxAsync(ctx, rpc, tx)
+			select {
+			case <-ctx.Done():
+				return
+			case txResults <- result:
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case txResultsToReport <- result:
+			}
+		}()
+	})
+	if err != nil {
+		primaryNodeWg.Wait()
+		close(txResultsToReport)
+		close(txResults)
+		return 0, err
+	}
+
+	// This needs to be done in parallel so the reporting knows when it's done (when the channel is closed)
+	txSender.wg.Add(1)
```

Comment:
I moved the `IfNotStopped` check inside of MultiNode `DoAll()` instead. The functionality is the same since `IfNotStopped` only checks the state once initially.